### ISSUE TITLE
RSA Testing Granularity

### ISSUE THREAD

**Issue by reaperhulk** [pyca/cryptography#1074]
Our current tests assume that if you support RSABackend you support PKCS1v1.5 (both encrypt/decrypt and sign/verify) as well as PSS (sign/verify) and OAEP (encrypt/decrypt). We do have an interface method called `mgf1_hash_supported` that allows us to query the backend to determine if the specific hash is supported for MGF1 purposes, but that is insufficiently expressive since OpenSSL allows any hash it knows for MGF1 with PSS but only allows SHA1 for OAEP MGF1 as of 1.0.1g. Additionally, it is assumed that the backend can handle (for both generation and loading) unusual bit lengths like 1029.

With the current work to expand the CommonCrypto backend to support asymmetric things those assumptions are no longer accurate. 

### Generation

CommonCrypto can load 1029 bit keys, but it can only generate a limited subset of key sizes (8-bit aligned at least, possibly more restrictive). The public exponent is also fixed at 65537.

### Signing/Verification

PKCS1v1.5 is supported. SHA1 & SHA2 are supported hash functions but this is a bit more nebulous (we could relatively easily create the ASN.1 wrapping for the hash function OID ourselves if it proves to be too inflexible). PSS is *not* supported and we currently have no way to query a backend to obtain this information.

### Encrypt/Decrypt

PKCS1v1.5 and OAEP are both supported. OAEP appears to support parameter/label as well as alternate MGF1 hash functions, but as previously noted our `mgf1_hash_supported` function may need to be expanded to know which padding (PSS or OAEP) is being requested so it can be used here. We do **not** have non-SHA1 MGF1 test vectors at this time.

### Solutions?

To fix this we could add `generate_rsa_parameters_supported`, `rsa_padding_supported`, and `rsa_oaep_mgf1_hash_supported` (and rename `mgf1_hash_supported` to `rsa_pss_mgf1_hash_supported`?) functions to the RSABackend interface. We would then need to refactor the tests significantly to use them for conditional skips. Once this is done the new commoncrypto work could potentially be landed.

Other ideas?

**Comment by reaperhulk** [comment]
I'm closing this as most of this has been addressed. We can open a new issue if we need to in the future.
